include/deal.II/base/tensor.h

```diff
@@ -106,62 +348,80 @@ class Tensor
   typedef typename numbers::NumberTraits<Number>::real_type real_type;
 
   /**
+   * Type of objects encapsulated by this container and returned by
+   * operator[](). This is a tensor of lower rank for a general tensor, and
+   * a scalar number type for Tensor<1,dim,Number>.
+   */
+  typedef typename Tensor<rank_-1,dim,Number>::tensor_type value_type;
+
+  /**
    * Declare an array type which can be used to initialize an object of this
    * type statically.
    */
-  typedef typename Tensor<rank_-1,dim,Number>::array_type array_type[dim];
+  typedef typename Tensor<rank_-1,dim,Number>::array_type
+  array_type[(dim != 0) ? dim : 1];
+  // ... avoid a compiler warning in case of dim == 0 and ensure that the
+  // array always has positive size.
 
   /**
-   * Constructor. Initialize all entries to zero.
+   * Constructor. Initialize all entries to zero if
+   * <tt>initialize==true</tt>; this is the default behaviour.
    */
-  Tensor ();
+  explicit
+  Tensor (const bool initialize = true);
```

Comment:
Yes, we could think about that but I don't know how feasible it is (because it all depends on the behavior of Tensor<0,dim>. In any case, I don't think it will be overly critical because compilers tend to get better in removing initialize-to-zero code in places where it would soon be overwritten. And I guess the soon-to-be-overwritten case it the only one where not initializing the array would make sense. The other relevant case I can imagine right now is a long std::vector/AlignedVector of Tensors but there one provides an initializer element already so better copy zero than undefined memory. :-)